programming_assignments/api.py

```diff
@@ -0,0 +1,433 @@
+    def list_assignments_stud(self, request, pk):
+        return self.list_assign_stud(request, pk)
+
+    @action(detail=True, methods=["GET"])
+    def retrieve_assignment(self, request, pk):
+        return self.retrieve_assign(request, pk)
+
+    @action(detail=True, methods=["PUT", "PATCH"])
+    def update_assignment(self, request, pk):
+        return self.update_assign(request, pk)
+
+    @action(detail=True, methods=["DELETE"])
+    def delete_assignment(self, request, pk):
+        return self.delete_assign(request, pk)
+
+
+class AssignmentSectionViewSet(
+    viewsets.GenericViewSet,
+    custom_mixins.InsOrTACreateMixin,
+    custom_mixins.UpdateMixin,
+    custom_mixins.DeleteMixin,
+):
+    """ViewSet for `AssignmentSectionViewSet()`."""
+
+    queryset = AssignmentSection.objects.all()
+    serializer_class = AssignmentSectionSerializer
+    permission_classes = (IsInstructorOrTA,)
+
+    @action(detail=False, methods=["POST"])
+    def create_section(self, request):
+        """Adds a section to the assignment.
+
+        Args:
+            request (Request): DRF `Request` object
+
+        Returns:
+            `Response` with the created section data and status `HTTP_201_CREATED`
+
+        Raises:
+            `HTTP_400_BAD_REQUEST`: Raised due to `create()` method
+            `HTTP_401_UNAUTHORIZED`: Raised by `IsInstructorOrTA` permission
+                class
+            `HTTP_403_FORBIDDEN`: Raised due to `create()` method
+            `HTTP_404_NOT_FOUND`: Raised due to `create()` method
+        """
+        assignment_id = request.data["assignment"]
+
+        try:
+            assignment = SimpleProgrammingAssignment.objects.select_related(
+                "course"
+            ).get(id=assignment_id)
+        except SimpleProgrammingAssignment.DoesNotExist as e:
+            logger.exception(e)
+            return Response(str(e), status.HTTP_404_NOT_FOUND)
+        course_id = assignment.course_id
+
+        return self.create(request, course_id)
+
+    @action(detail=True, methods=["GET"])
+    def list_sections(self, request, pk):
+        """Gets all the section in the current assignment.
+
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): SimpleProgrammingAssignment id
+
+        Returns:
+            `Response` with all the section data and status `HTTP_200_OK`
+
+        Raises:
+            `HTTP_401_UNAUTHORIZED`: Raised by `IsInstructorOrTA` permission class
+            `HTTP_403_FORBIDDEN`: Raised due to `list()` method
+            `HTTP_404_NOT_FOUND`: Raised due to `list()` method
+        """
+        user = request.user
+        try:
+            assignment = SimpleProgrammingAssignment.objects.get(id=pk)
+        except SimpleProgrammingAssignment.DoesNotExist as e:
+            logger.exception(e)
+            return Response(str(e), status.HTTP_404_NOT_FOUND)
+
+        course_id = assignment.course_id
+        role = CourseHistory.objects.get(
```

Comment:
consider a  scenario when user has account but not enrolled in course.... then he can access all sections